implementation of export button for submissions

Balgorn opened the pull request.

krusche commented:
Please get the newest version from develop and reimplement the client functionality using TypeScript

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -233,6 +233,65 @@ public void cleanup(Long id, boolean deleteRepositories) throws IOException {
     }
 
+    @Transactional(readOnly = true)
+    public java.io.File exportParticipations(Long id, List<String> studentIds) {
```
- please add JavaDoc comments for all public methods
- What kind of id do we expect here? Would be better to call it participationId

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -233,6 +233,65 @@ public void cleanup(Long id, boolean deleteRepositories) throws IOException {
+                        zippedRepoFiles.add(zippedRepoFile);
+                        boolean allowInlineEditor = ((ProgrammingExercise) exercise).isAllowOnlineEditor() != null && ((ProgrammingExercise) exercise).isAllowOnlineEditor();
+                        if(!allowInlineEditor){ //if onlineeditor is not allowed we are free to delete
+                            log.info("Delete temporary repoistory "+ repo.getLocalPath().toString());
```
please change all logs to debug level, so that they don't pollute the logs on the production server

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-repo-export-dialog.component.ts:
```diff
@@ -0,0 +1,86 @@
+export class InstructorDashboardExportReposComponent {
+
+    exercise: Exercise;
+    confirmExerciseName;
```
please use types for all variables

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-repo-export-dialog.component.ts:
```diff
@@ -0,0 +1,86 @@
+
+    exportRepos(id: number) {
+        this.exportInProgress = true;
+        const list = this.list.split(',').map( e => e.trim());
```
improve variable names, what kind of list are you using here?

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-repo-export-dialog.component.ts:
```diff
@@ -0,0 +1,86 @@
+        this.activeModal.dismiss('cancel');
+    }
+
+    exportRepos(id: number) {
```
improve variable names, what kind of id are you using here?

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-repo-export-dialog.component.ts:
```diff
@@ -0,0 +1,86 @@
+              this.activeModal.dismiss(true);
+              this.exportInProgress = false;
+              if (response.body) {
+                const blob = new Blob([response.body], { type: 'application/zip' });
```
blob is not a good variable name

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-repo-export-dialog.component.ts:
```diff
@@ -0,0 +1,86 @@
+})
+export class InstructorDashboardExportReposPopupComponent implements OnInit, OnDestroy {
+
+    routeSub: any;
```
can we use a proper type and avoid any?

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:
```diff
@@ -327,5 +328,40 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+        List<String> studentList = Arrays.asList(studentIds.split("\\s*,\\s*"));
+
```
What happens if studentIds is null or empty? Could we catch this case earlier and return then to avoid any kind of problem that could arise because of it and to improve performance?

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-archive-dialog.component.html:
```diff
@@ -1,13 +1,13 @@
 <form name="archiveForm" role="form" (ngSubmit)="confirmArchive(exercise.id)" #archiveForm="ngForm">
     <div class="modal-header">
-        <h4 class="modal-title" jhiTranslate="entity.archive.title">Confirm archive operation</h4>
+        <h4 class="modal-title" jhiTranslate="entity.exportRepos.warning">Confirm Export</h4>
```
I think this is a mistake and you accidentally changed this file. Please revert it.

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard-archive-dialog.component.html:
```diff
@@ -1,13 +1,13 @@
         <jhi-alert-error></jhi-alert-error>
         <p jhiTranslate="instructorDashboard.archive.question"
-           translateValues="{exerciseTitle: '{{exercise.title}}', courseTitle: '{{exercise.course.title}}'}">Confirm archive
+           translateValues="{exerciseTitle: '{{exercise.title}}', courseTitle: '{{exercise.course.title}}'}">Enter all ids of students you want to download the repo from
```
I think this is a mistake and you accidentally changed this file. Please revert it.

krusche commented on src/main/webapp/i18n/en/global.json:
```diff
@@ -128,6 +128,7 @@
             "addimage": "Add image",
             "back": "Back",
             "cancel": "Cancel",
+            "exportRepos": "Download",
```
Download Repositories

krusche commented on src/main/webapp/i18n/en/instructorDashboard.json:
```diff
@@ -13,6 +13,11 @@
+        "exportRepos": {
+            "warning": "Confirm Export",
+            "title": "Download Repos",
+            "question": "Enter all ids of students you want to download the repo from"
```
Can you write "logins" instead of "ids" and give a simple example for TUM students, e.g. 'ga87fix, ne23kow'

Balgorn commented:
@krusche 

- reverted back the archive dashboard, that was an accident sorry for that.
- changed the wording for the repository download
- made the splitting of student logins safer

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -233,6 +233,72 @@ public void cleanup(Long id, boolean deleteRepositories) throws IOException {
+     * @return a zip file containing all requested participations
+     */
+    @Transactional(readOnly = true)
+    public java.io.File exportParticipations(Long id, List<String> studentIds) {
```
It would be better to rename 'id' to 'exerciseId'

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:
```diff
@@ -327,5 +328,43 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
+    @GetMapping(value = "/exercises/{id}/participations/{studentIds}")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    @Timed
+    public ResponseEntity<Resource> exportSubmissions(@PathVariable Long id, @PathVariable String studentIds) throws IOException {
```
Please rename 'id' to 'exerciseId'

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:
```diff
@@ -327,5 +328,43 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
+        }
+        List<String> studentList = Arrays.asList(studentIds.split("\\s*,\\s*"));
+        if(studentList.isEmpty() || studentList == null){
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
Can we provide more feedback here why the request was bad?

krusche commented on src/main/webapp/app/instructor-dashboard/instructor-dashboard.component.html:
```diff
@@ -31,6 +31,14 @@ <h2>{{exercise.course.title}} - {{exercise.title}} <small>{{results.length}} res
             <button class="btn btn-primary btn-sm" (click)="exportResults()">
                 <i class="fa fa-download fa-fw"></i>&nbsp;Export Results
             </button>
+            <button *ngIf="exercise.type !== 'quiz'"
```
we also have modeling exercises, I think we need to exclude them here as well, because the zip export makes only sense for programming-exercises. As we will add more exercises in the future, I suggest, we explicitly test for programming exercises here

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:
```diff
@@ -327,5 +328,43 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
+        * @param studentIds the studentIds seperated via semicolon to get their submissions
+     * @return ResponseEntity with status
+     */
+    @GetMapping(value = "/exercises/{id}/participations/{studentIds}")
```
What happens, if I have any whitespaces in the studentIds? Can it still be represented as URL?

Balgorn replied:
the front end actually filters them out. In addition I also added this replacement to the server in case of malformed requests. In addition I also tested positively that all involved method calls handle them correctly.
